// fast-json-parser/lib/tests/fuzz.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function int(start, end) {
    return start + Math.floor(Math.random() * (end + 0.999999));
}
exports.int = int;
function number() {
    var v = Math.random();
    if (v < 0.1) {
        v = int(-3, 3);
    }
    else {
        return (Math.random() - 0.5) * 100000;
    }
}
exports.number = number;
var sTable = '\\"`1234567890!@#$::\\""\'\'qwertyuiopasdfghjklzxcvbnm\r\n\t\a';
function string() {
    var n = int(0, 20) * int(0, 5), s = '';
    for (var k = 0; k < n; ++k) {
        s += sTable[int(0, sTable.length - 1)];
    }
    return s;
}
exports.string = string;
var others = [null, true, false];
function other() {
    return others[int(0, 2)];
}
exports.other = other;
function array() {
    var n = int(0, 20), s = [];
    for (var k = 0; k < n; ++k) {
        s.push(value());
    }
    return s;
}
exports.array = array;
function object() {
    var o = {};
    var n = int(0, 20);
    for (var k = 0; k < n; ++k) {
        o[string()] = value();
    }
    return o;
}
exports.object = object;
function value() {
    var choice = Math.random();
    if (choice < 0.3)
        return number();
    if (choice < 0.6)
        return string();
    if (choice < 0.92)
        return other();
    if (choice < 0.96)
        return array();
    return object();
}
exports.value = value;
